psycopg/psycopg/_queries.py

```diff
@@ -96,6 +95,107 @@ def dump(self, vars: Optional[Params]) -> None:
             self.types = ()
             self.formats = None
 
+    @staticmethod
+    def is_params_sequence(vars: Params) -> bool:
+        # Try concrete types, then abstract types
+        t = type(vars)
+        if t is list or t is tuple:
+            sequence = True
+        elif t is dict:
+            sequence = False
+        elif isinstance(vars, Sequence) and not isinstance(vars, (bytes, str)):
+            sequence = True
+        elif isinstance(vars, Mapping):
+            sequence = False
+        else:
+            raise TypeError(
+                "query parameters should be a sequence or a mapping,"
+                f" got {type(vars).__name__}"
+            )
+        return sequence
+
+    @staticmethod
+    def validate_and_reorder_params(
+        parts: List[QueryPart], vars: Params, order: Optional[List[str]]
+    ) -> Sequence[Any]:
+        """
+        Verify the compatibility between a query and a set of params.
+        """
+        sequence = PostgresQuery.is_params_sequence(vars)
+
+        if sequence:
+            if len(vars) != len(parts) - 1:
+                raise e.ProgrammingError(
+                    f"the query has {len(parts) - 1} placeholders but"
+                    f" {len(vars)} parameters were passed"
+                )
+            if vars and not isinstance(parts[0].item, int):
+                raise TypeError("named placeholders require a mapping of parameters")
+            return vars  # type: ignore[return-value]
+
+        else:
+            if vars and len(parts) > 1 and not isinstance(parts[0][1], str):
+                raise TypeError(
+                    "positional placeholders (%s) require a sequence of parameters"
+                )
+            try:
+                return [vars[item] for item in order or ()]  # type: ignore[call-overload]
+            except KeyError:
+                raise e.ProgrammingError(
+                    "query parameter missing:"
+                    f" {', '.join(sorted(i for i in order or () if i not in vars))}"
+                )
+
+    @staticmethod
+    def query2pg(
+        query: bytes, encoding: str, vars: Params
+    ) -> Tuple[bytes, Optional[List[PyFormat]], Optional[List[str]], List[QueryPart]]:
```

Comment:
This has now lost the `@lru_cache`, which is very important.

The need to have the cache was the reason why the function doesn't take the `vars`. They are unused in this method, but they make it impossible to cache.

Reply:
Sorry I should have mentioned I left that out because I couldn't figure out the unhashability of list error, but now I've fixed it, which is a better solution anyway, since passing the number of parameters is all we need. See new commit.